### DISCUSSION TITLE
Why does the new version get the state before taking the action?

### DISCUSSION THREAD

**Discussion by jacobtw2** [MarcEscandell/ALPypeRL#8] [General]
Hi,

I was building a model with ALpypeRL and I noticed that the new release gets the observation before taking the action in the sim and then returns that observation at the end of the step method

Why has it been done this way?

Usually you want to return the state the simulation is in after the  step has been performed

**Comment by MarcEscandell** [comment]
Hei Jacob! Thanks for reaching out.

You are correct. The cycle followed by _ALPypeRL_ is:
1. Get **observation**
2. Apply **action**
3. Get **reward**

> Why has it been done this way?

I had this question myself during design. After consulting multiple resources, decided to stick with it. Would you mind sharing and example or resource that supports your understanding? I'll be happy to change it then.

**Comment by jacobtw2** [comment]
https://gymnasium.farama.org/api/env/#gymnasium.Env.step

Here is a link to gymnasium.Env which is used for Alpyperl. If you look at the env.step() method description it says:

RETURNS:
observation (ObsType) – An element of the environment’s [observation_space](https://gymnasium.farama.org/api/vector/#observation_space) as the next observation due to the agent actions. An example is a numpy array containing the positions and velocities of the pole in CartPole.

This is saying that it should return the observation space which is obtained after the action is taken

**Reply by MarcEscandell** [reply]
Hei @jacobtw2 !
Thanks for sharing! I have also reviewed one of the _classic control_ examples implementation, the **cartpole**, under `gymnasium/envs/classic_control/cartpole.py` (link [here](https://github.com/Farama-Foundation/Gymnasium/blob/main/gymnasium/envs/classic_control/cartpole.py)). This should be quite reliable. In the code, it also appears to execute the action first before retrieving the observation and reward.

Here's a snippet:

```python
    def step(self, action):
        assert self.action_space.contains(
            action
        ), f"{action!r} ({type(action)}) invalid"
        assert self.state is not None, "Call reset before using step method."
        # =======================================================
        # Execution of the "action" starts here ------>
        # =======================================================
        x, x_dot, theta, theta_dot = self.state
        force = self.force_mag if action == 1 else -self.force_mag
        costheta = math.cos(theta)
        sintheta = math.sin(theta)

        # For the interested reader:
        # https://coneural.org/florian/papers/05_cart_pole.pdf
        temp = (
            force + self.polemass_length * theta_dot**2 * sintheta
        ) / self.total_mass
        thetaacc = (self.gravity * sintheta - costheta * temp) / (
            self.length * (4.0 / 3.0 - self.masspole * costheta**2 / self.total_mass)
        )
        xacc = temp - self.polemass_length * thetaacc * costheta / self.total_mass

        # =======================================================    
        # Retrieval of the "observation" is done here  ------>
        # =======================================================
        if self.kinematics_integrator == "euler":
            x = x + self.tau * x_dot
            x_dot = x_dot + self.tau * xacc
            theta = theta + self.tau * theta_dot
            theta_dot = theta_dot + self.tau * thetaacc
        else:  # semi-implicit euler
            x_dot = x_dot + self.tau * xacc
            x = x + self.tau * x_dot
            theta_dot = theta_dot + self.tau * thetaacc
            theta = theta + self.tau * theta_dot

        self.state = (x, x_dot, theta, theta_dot)

        terminated = bool(
            x < -self.x_threshold
            or x > self.x_threshold
            or theta < -self.theta_threshold_radians
            or theta > self.theta_threshold_radians
        )

        # =======================================================    
        # Calculation of the "reward" is done here  ------>
        # =======================================================
        if not terminated:
            if self._sutton_barto_reward:
                reward = 0.0
            elif not self._sutton_barto_reward:
                reward = 1.0
        elif self.steps_beyond_terminated is None:
            # Pole just fell!
            self.steps_beyond_terminated = 0
            if self._sutton_barto_reward:
                reward = -1.0
            elif not self._sutton_barto_reward:
                reward = 1.0
        else:
            if self.steps_beyond_terminated == 0:
                logger.warn(
                    "You are calling 'step()' even though this "
                    "environment has already returned terminated = True. You "
                    "should always call 'reset()' once you receive 'terminated = "
                    "True' -- any further steps are undefined behavior."
                )
            self.steps_beyond_terminated += 1
            if self._sutton_barto_reward:
                reward = -1.0
            elif not self._sutton_barto_reward:
                reward = 0.0

        if self.render_mode == "human":
            self.render()
        # truncation=False as the time limit is handled by the `TimeLimit` wrapper added during `make`
        return np.array(self.state, dtype=np.float32), reward, terminated, False, {}
```

I'll apply the changes as soon as possible. In fact, changing the order might fix one of the requests you had in the past: to avoid "early retrieval of observation" or "calling `getObservation` in your AnyLogic model too early".

**Reply by MarcEscandell** [reply]
Hei Jacob! Sorry for picking this up so late.

I did a bit more research on this topic. The general rule for an RL loop is to **get observation**, then **take action** and, finally, **collect the reward**. However, in the piece of code shared earlier from `gymnasium`, _action_ and _observation_ seemed to be reversed. If I am not mistaken this is because the observation is not being pared with the action in the same step, but in the next one. That is why the model, first of all, will call the `reset()` function, which at the same time returns the first observation.

Additional work that I have done in this area is to compare the performance with the old and the new implementation. I seem to see, when using the _CartPole_ model, that performance improves (at least in terms of reward improvement rate) when using the new arrangement (what you suggested).

Here's a plot comparison:
![image](https://github.com/user-attachments/assets/d9c05bf1-0fd6-4599-aa22-990d6528ff47)

* Blue line is the new implementation (_take action_ then _get observation_)
* Orange line is the old implementation (_get observation_ then _take action_)

**Reply by MarcEscandell** [reply]
Please use version **alpyperl-1.1.0**!
